rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,82 @@
+
+The stable version should be updated after each Cloud Tenant upgrade. This should follow the same workflow as updating the cloud-stable agent version endpoint. The version should be updated soon after Cloud Tenant upgrades have been completed.
+
+Unlike the agent version discovery model, there will not be an endpoint to identify if a critical update is available. The reason for this is because updates are not executed in a scheduled update window. Teleport client tools will check this endpoint on every login and update accordingly if a new version is available.
+
+On every login, tsh will pull the current cloud-stable version of Teleport and store the value in the .tsh directory under `$HOME/.tsh/cloud-stable-version`.
+
+### Caching
+
+A single version of the Teleport client tools will be stored in the user's $HOME directory. Teleport already makes use of a .tsh directory for storing tsh config. Cached versions of tsh/tctl will also live within this directory under ` $HOME/.tsh/cache/<version>/{tsh,tctl}`. Whenever a new version of the client tools are available, the existing cached version will be replaced.
+
+On every login, tsh will compare the currently cached version of the client tools with the current stable version stored in `$HOME/.tsh/cloud-stable-version`. If the cached version is behind the current stable version, a new version of the client tools will be downloaded to replace the existing ones. When downloading the Teleport client tools, the binaries will be downloaded directly. We need to make sure to download a version that is compatible with the system.
+
+Because only a single version of the client tools will be cached at a time, we do not need to worry about cleaning up the cache.
+
+### Package upgrade
+
+As an alternative to caching, we've considered directly upgrading the Teleport packages.
+
+This approach will be difficult to implement because we support multiple package repositories and installation methods. The amount of maintenance and testing required to support this solution is probably not worth the effort at this time.
+
+Another downside to this approach would be that upgrades would require sudo/admin privileges.
+
+### Config
+
+By default, the client tools will have auto-updates enabled. I don't see compelling reason to allow users to disable auto updates for client tools. But if we want to allow this feature to be disabled, we can add support in the tsh configuration files by adding a DISABLE_AUTO_UPDATES flag.
+
+## UX
+
+Ideally, we'd like this feature to be integrated seamlessly. Users of the Teleport client tools should not need to take any actions to enable auto updates for their client tools.
+
+We should be transparent about the client tools updates. Whenever a new cloud-stable version is available for download, the progress of the update will be output to stdout. e.g. `New cloud-stable version of Teleport detected`, `Downloading latest version of tsh/tctl... `, `Updated tsh/tctl to version v13.2.3!`.
```

Comment:
Should we present some confirmation? @xinding33 What do you think?

Reply:
For CLIs, I typically see asking for confirmation. I know @klizhentas doesn't like adding knobs, but I think it's worth it here. IMO, the UX should be along the following lines:

- Upon first available update opportunity, client will ask user two questions: 1) if they would like to update to the latest version (default yes, meaning if they simply click `enter` they respond with yes) and 2) if they would like to enroll in automatic updates (again, default yes)
- User can opt out of automatic updates by changing some config option
- If user opts out, we still prompt them with the first question every time an update is available but not the second question (however we should display a message every time suggesting that they enable automatic updates)

Reply:
Sounds good to me.

Reply:
It's an edge case but confirmation prompts tend to break CIs. Not sure if there are any use cases that pertain to us here but thought I'd bring it up just in case.

Reply:
my 2c:
1. add a flag for this like `--[no]-auto-update` (or something)
2. without the flag given, when a tty is detected and an update is available we prompt for a user decision
3. without the flag given and not a tty, default to not auto-updating so scripts are not broken